libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,955 @@
+        debug(" WARNING - %s HW: %s",TAG,"EMPTY WIRE");
+        return;
+    }
+
+    debug("%80s %d", "READ RESPONSE incoming_bytes_count:",_incoming_bytes_count);
+    for (int i = 0; i < _incoming_bytes_count; ++i) {
+        uint8_t _response_in_bytes = _port->read();
+        debug(" READ RESPONSE _response_in_bytes:%02X",_response_in_bytes);
+        if ((inline_buffer_length == 0 && _response_in_bytes != SMARTAUDIO_SYNC_BYTE)
+            || (inline_buffer_length == 1 && _response_in_bytes != SMARTAUDIO_HEADER_BYTE)) {
+            debug(" READ RESPONSE byte discard:%02X",_response_in_bytes);
+            inline_buffer_length = 0;
+        } else if (inline_buffer_length < 16) {
+            response_buffer[inline_buffer_length++] = _response_in_bytes;
+            debug(" READ RESPONSE byte -> response_buffer added :%02X",response_buffer[inline_buffer_length-1]);
+        }
+    }
+
+    //last_io_time = AP_HAL::millis();
+
+    if (inline_buffer_length < _response_header_size) {
+        debug("INPUT FILTERED FROM UART IS SHORTER THAN EXPLICIT RESPONSE LENGTH");
+        is_waiting_response=false;
+        return;
+    }
+
+    uint8_t payload_len = response_buffer[3];
+
+    if (inline_buffer_length < payload_len + _response_header_size) {
+        // Not all response bytes received yet, splitted response is incoming
+        debug("RESPONSE SEEMS NOT TO BE COMPLETE WAITING IN OTHER LOOP CYCLE ?");
+        return;
+    } else {
+        inline_buffer_length = payload_len + _response_header_size;
+    }
+    is_waiting_response=false;
+   
+    debug(" PARSING FRAME RESPONSE");
+    parse_frame_response(response_buffer);
+    debug(" PARSING FRAME RESPONSE ENDED");
+ 
+    response_buffer=nullptr;
+}
+
+/**
+ * This method parse the frame response and match the current device configuration  *
+ *
+ * This method only changes the internal state vtx buffer. Using response parser and then updating the internal buffer.
+ * */
+bool AP_SmartAudio::parse_frame_response(const uint8_t *buffer)
+{
+    if (buffer!=nullptr) {
+
+        smartaudioSettings_t _vtx_settings;
+        // process response buffer
+        if(!smartaudioParseResponseBuffer(&_vtx_settings,buffer)){
+            debug("%80s","Unparseable buffer response");
+            return false;
+        }
+
+        debug("%80s %02X","selecting update path using:",_vtx_settings.update_flags);
+        // if partial updates because setters method to vtx.
+        if(_vtx_settings.update_flags!=0x0F){ 
+            debug("%80s","parse_frame_response:: overall update flow path");
+
+            smartaudioSettings_t _current_vtx_settings;
+
+            // take the current vtx info
+            vtx_states_queue.peek(&_current_vtx_settings,1);
+
+            // get update flags from the parsed response_buffer
+            _current_vtx_settings.update_flags=_vtx_settings.update_flags;
+
+
+            if( _vtx_settings.update_flags & (1 << 0)){ // freq has changed
```

Comment:
There should be an enum for these values - might even be one AP_VideoTX already

Reply:
Fixed. 

        // Proposed to be into AP_VideoTX
        enum class HWVtxUpdates {
            OVERALL_UPD = 0X0F,
            FREQ_UPD = 1 << 0,
            CHAN_UPD = 1 << 1,
            POW_UPD  =  1 << 2,
            MODE_UPD =  1 << 3,
            NO_UPD   = 0
        };

Reply:
Fixed in 8d208cdac75896b4404257600404fb1403fa4e8f